File: app/src/main/java/mm/pndaza/tipitakamyanmar/adapter/BookmarkAdapter.java
package mm.pndaza.tipitakamyanmar.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.RecyclerView;

import com.sdsmdg.tastytoast.TastyToast;

import java.util.ArrayList;

import mm.pndaza.tipitakamyanmar.R;
import mm.pndaza.tipitakamyanmar.db.DBOpenHelper;
import mm.pndaza.tipitakamyanmar.model.Bookmark;
import mm.pndaza.tipitakamyanmar.utils.MDetect;
import mm.pndaza.tipitakamyanmar.utils.MyanNumber;
import mm.pndaza.tipitakamyanmar.utils.Rabbit;


public class BookmarkAdapter extends RecyclerView.Adapter<BookmarkAdapter.ViewHolder> {

    private ArrayList<Bookmark> bookmarkList;
    private View.OnClickListener onClickListener;

    private Context context;

    public BookmarkAdapter(ArrayList<Bookmark> bookmarkList) {
        this.bookmarkList = bookmarkList;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        context = parent.getContext();
        LayoutInflater inflater = LayoutInflater.from(context);
        // reuse layout because layout are same
        View wordListItemView = inflater.inflate(R.layout.bookmarklist_row_item, parent, false);
        return new ViewHolder(wordListItemView);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        final Bookmark bookmark = bookmarkList.get(position);
        holder.tvNote.setText(bookmark.getNote()); // note are saved as device encoding
        holder.tvBookName.setText(MDetect.getDeviceEncodedText(bookmark.getBookName()));
        String pageNumber = MDetect.getDeviceEncodedText("နှာ - ") + MyanNumber.toMyanmar(bookmark.getPageNumber());
        holder.tvPageNumber.setText(pageNumber);
    }

    @Override
    public int getItemCount() {
        return bookmarkList.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        TextView tvNote;
        TextView tvBookName;
        TextView tvPageNumber;

        ViewHolder(@NonNull View itemView) {
            super(itemView);

            tvNote = itemView.findViewById(R.id.tv_note);
            tvBookName = itemView.findViewById(R.id.tv_bookName);
            tvPageNumber = itemView.findViewById(R.id.tv_pageNumber);

            itemView.setTag(this);
            itemView.setOnClickListener(onClickListener);
        }
    }

    public void setOnClickListener(View.OnClickListener clickListener) {
        onClickListener = clickListener;
    }


    public void deleteItem(int position){

        AlertDialog.Builder alertDialog = new AlertDialog.Builder(context,R.style.AlertDialogTheme);
        String message = "သိမ်းမှတ်ထားသည်ကို ဖျက်မှာလား";
        String comfirm = "ဖျက်မယ်";
        String cancel = "မလုပ်တော့ဘူး";
        if (!MDetect.isUnicode()) {
            message = Rabbit.uni2zg(message);
            comfirm = Rabbit.uni2zg(comfirm);
            cancel = Rabbit.uni2zg(cancel);
        }

        alertDialog.setMessage(message)
                .setCancelable(true)
                .setPositiveButton(comfirm,
                        (dialog, id) -> {
                            DBOpenHelper.getInstance(context).removeFromBookmark(position);
                            bookmarkList.remove(position);
                            notifyDataSetChanged();
                            TastyToast.makeText(context,
                                    MDetect.getDeviceEncodedText("သိမ်းထားသည်ကို ဖျက်လိုက်ပြီးပါပြီ"),
                                    TastyToast.LENGTH_SHORT, TastyToast.INFO).show();
                        })
                .setNegativeButton(cancel, (dialog, id) -> {
                    notifyDataSetChanged();
                });
        alertDialog.show();
    }

}

File: app/src/main/java/mm/pndaza/tipitakamyanmar/adapter/BookmarkAdapter.kt
package mm.pndaza.tipitakamyanmar.adapter

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import mm.pndaza.tipitakamyanmar.databinding.BookmarklistRowItemBinding
import mm.pndaza.tipitakamyanmar.model.Bookmark
import mm.pndaza.tipitakamyanmar.utils.MDetect
import mm.pndaza.tipitakamyanmar.utils.NumberUtil

class BookmarkAdapter(
    private var bookmarks: MutableList<Bookmark>,
    private val onItemClick: (Bookmark) -> Unit,
) : RecyclerView.Adapter<BookmarkAdapter.ViewHolder>() {

    inner class ViewHolder(val binding: BookmarklistRowItemBinding) :
        RecyclerView.ViewHolder(binding.root)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = BookmarklistRowItemBinding.inflate(
            LayoutInflater.from(parent.context), parent, false
        )
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val bookmark = bookmarks[position]
        with(holder.binding) {
            tvNote.text = bookmark.note
            tvBookName.text = MDetect.getInstance().getDeviceEncodedText(bookmark.bookName)
            val pageNumber = MDetect.getInstance().getDeviceEncodedText("နှာ - ") +
                    NumberUtil.toMyanmar(bookmark.pageNumber)
            tvPageNumber.text = pageNumber

            root.setOnClickListener { onItemClick(bookmark) }
        }
    }

    override fun getItemCount() = bookmarks.size

    fun getBookmarkAt(position: Int): Bookmark? {
        return bookmarks.getOrNull(position)
    }

    fun removeItemAt(position: Int): Bookmark? {
        if (position in bookmarks.indices) {
            val removeItem = bookmarks.removeAt(position)
            notifyItemRemoved(position)
            return removeItem
        }
        return null
    }

    fun restoreItemAt(bookmark: Bookmark, position: Int) {
        if (position in 0..bookmarks.size) {
            bookmarks.add(position, bookmark)
            notifyItemInserted(position)
        }
    }

    /*
    fun updateBookmarks(newBookmarks: List<Bookmark>) {
        val oldSize = bookmarks.size
        bookmarks.clear()
        bookmarks.addAll(newBookmarks)
        notifyItemRangeChanged(0, maxOf(oldSize, bookmarks.size))
    }
    */

    // Update data using DiffUtil for better performance and animations
    fun submitList(newBookmarks: List<Bookmark>) {
        val diffCallback = BookmarkDiffCallback(this.bookmarks, newBookmarks)
        val diffResult = DiffUtil.calculateDiff(diffCallback)
        // Important: Update the internal list *after* calculating diff
        this.bookmarks.clear()
        this.bookmarks.addAll(newBookmarks)
        diffResult.dispatchUpdatesTo(this)
    }

    // DiffUtil Callback implementation
    private class BookmarkDiffCallback(
        private val oldList: List<Bookmark>,
        private val newList: List<Bookmark>
    ) : DiffUtil.Callback() {

        override fun getOldListSize(): Int = oldList.size
        override fun getNewListSize(): Int = newList.size

        // Check if items represent the same object (unique ID)
        override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            return oldList[oldItemPosition].rowId == newList[newItemPosition].rowId
        }

        // Check if item contents are the same (if IDs match)
        @SuppressLint("DiffUtilEquals") // Using data class equals is usually sufficient
        override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            return oldList[oldItemPosition] == newList[newItemPosition]
        }
    }
}